### PR TITLE
prov/verbs: add MSG capability for EP_RDM

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
This series adds MSG capability for EP_RDM and fixes for newly run fabtests

new fabtests  passes:

- fi_rdm -f verbs
- fi_rdm_pingpong -I 5 -f verbs
- fi_rdm_tagged_peek -f verbs
- fi_rdm_rma -o write -I 5 -f verbs
- fi_rdm_rma -o read -I 5 -f verbs

new fabtests  fails:

- fi_rdm_shared_av -f verbs

@a-ilango , please review

### CODE REVIEW

File: prov/verbs/src/verbs_info.c [hunk 1]
```diff
@@ -361,7 +361,8 @@ int fi_ibv_check_rx_attr(const struct fi_rx_attr *attr,
 		return -FI_ENODATA;
 	}
 
-	if (attr->total_buffered_recv > info->rx_attr->total_buffered_recv) {
+	if (!attr->total_buffered_recv &&
+		(attr->total_buffered_recv > info->rx_attr->total_buffered_recv)) {
```
Comment (a-ilango) [inline prov/verbs/src/verbs_info.c]: This seems incorrect. The second condition is evaluated only if attr->total_buffered_recv is zero.

**Comment by a-ilango** [comment]
The PR looks mostly good to me except for the one comment.

File: prov/verbs/src/verbs_info.c (hunk 1 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/verbs_info.c]: From man page: 

> Defines the total available space allocated by the provider to buffer messages that are received for which there is no matching receive operation. If set to 0, and the domain does not support FI_RM_ENABLED, any messages that arrive before a receive buffer has been posted are lost. When the domain supports FI_RM_ENABLED, the actual amount of buffering provided may exceed the value specified in total_buffered_recv.

The verbs/RDM is able to receive such many unexpected (without preposting) messages as system memory allows. As I understand, this means the provider does not support FI_RM_ENABLED and attr->total_buffered_recv should be set = 0, otherwise we need do second check for EP_MSG

Comment (a-ilango) [inline prov/verbs/src/verbs_info.c]: I'm not sure if I understand your comment. If a provider is able to receive unexpected messages without losing them, then that means the provider supports FI_RM_ENABLED. The app can choose to disable RM if it wants.

If the app requests a total_buffered_recv > available buffering, we have to return an error. The change in the PR allows this check only when total_buffered_recv is 0.

Comment (evgeny-leksikov) [inline prov/verbs/src/verbs_info.c]: @a-ilango thanks for clarification! I added FI_RM_ENABLED for EP_RDM and corresponding check to total_buffered_recv comparison

**Comment by evgeny-leksikov** [comment]
@a-ilango thanks for good catch! Now, that should be correct.

File: prov/verbs/src/verbs_info.c [hunk 2]
```diff
@@ -361,7 +362,12 @@ int fi_ibv_check_rx_attr(const struct fi_rx_attr *attr,
 		return -FI_ENODATA;
 	}
 
-	if (attr->total_buffered_recv > info->rx_attr->total_buffered_recv) {
+	rm_enabled =(info->domain_attr &&
+		     info->domain_attr->resource_mgmt == FI_RM_ENABLED);
+
+	if (!rm_enabled &&
+	    (attr->total_buffered_recv > info->rx_attr->total_buffered_recv))
+	{
```
Comment (a-ilango) [inline prov/verbs/src/verbs_info.c]: I believe the above check should be there even for FI_RM_ENABLED. If the user requests buffering of length > supported by the provider we have to return an error. In case of FI_RM_DISABLED, the provider's total_buffered_recv would always be zero. Sorry if I wasn't clear before. :)

So the original code before this PR's change was already correct. If verbs/RDM supports resource management, it should set the total_buffered_recv to the amount of buffering provided.

Comment (evgeny-leksikov) [inline prov/verbs/src/verbs_info.c]: hmm... let me do another iteration with understanding of the description :)

> When the domain supports FI_RM_ENABLED, the actual amount of buffering provided may exceed the value specified in total_buffered_recv.

So, this check has no meaning if FI_RM_ENABLED.

ofcause, I can set total_buffered_recv to -1 and the check may stay without modification, but I think updated patch does exactly what's described in the doc and matches to  real verbs/rdm behaviour, doesn't it?

Comment (a-ilango) [inline prov/verbs/src/verbs_info.c]: Okay, I think I'm understanding your point. It seems the man page statement could be interpreted as total_buffered_recv value being meaningless when domain supports FI_RM_ENABLED since the actual buffering provided might be higher.

I thought total_buffered_recv was the minimum value of the buffering supported by the provider and if the app requests something higher, it has to be rejected. Now I'm getting doubts as to whether this is an input parameter (app requesting the provider to allocate required buffering) or output parameter (provider specifying the amount of buffering supported). 

@shefty Can you clarify the man page description? I hope you notice the email notification for this comment in the mountain of emails that you're sifting through. :)

Comment (shefty) [inline prov/verbs/src/verbs_info.c]: The value is an in/out parameter.  It is independent from FI_RM_ENABLED.

When FI_RM_ENABLED is set to 1, sent data transfers will not be lost (baring any fatal error).  This is true whether or not there is extra buffering on the receive side (i.e. total_buffered_recv > 0).

If FI_RM_ENABLED is set to 0, then any message that overruns the receive side buffering will be lost.  This is true for any value of total_buffered_recv.  The man pages is simply calling out that if total_buffered_recv = 0, then the very first message that doesn't match with a user buffer will be lost.

Total_buffered_recv is an in/out parameter.  The comment that the amount of buffering may be greater just means that the provider is not restricted by this limit.  For example, if the provider implements a rendezvous protocol as part of its unexpected message handing, then it could _appear_ to the application that it buffers more data than what's specified for total_buffered_recv.  However, this is provider specific and the app cannot rely on this.

Hopefully this makes sense.  If there's any wording change that you'd like to the man page to clarify things, please let me know.

Comment (evgeny-leksikov) [inline prov/verbs/src/verbs_info.c]: @shefty Is total_buffered_recv a size of a single unexpected message or summarized size of all unexpected messages which a provider is able to process? The note about rendezvous protocol hints at the first case but what if a provider has possibility only for a limited number of unexpected messages, including RTS? Is there some another attribute?

**Comment by shefty** [comment]
It's the total amount of buffer space that's available.  Nothing beyond that is defined.  This issue has been brought up (and may be being tracked).  An app doesn't know if this means that a single message can consume the entire buffer, or if only, say, the first MTU sized bytes of each message will be cached.

For example, if total_buffered_recv is 8k, the current documentation doesn't say if that means a single 8k message will consume the entire buffer, or if the provider can actually buffer the first 2k of 4 different messages.  We need to be careful defining this too closely as it may force an implementation.   IMO, either option is valid.  I hesitate to add more attributes around how unexpected messages are handled.

**Comment by evgeny-leksikov** [comment]
Sorry, but why (and how) we should provide to user some value which he can't control?
And regarding to this PR, I'm confused what I should return if verb provider does not control buffering at all? All unexpected eager messages (including RTS) will be processed correctly while system memory is available otherwise FI_ENOMEM will be returned, huge messages will be synchronized by rndv protocol.

**Comment by shefty** [comment]
The user can control this.  It's just not a hard limit.  The verbs provider is using system memory for receive side buffering.  However, it's possible that other providers may need to post buffers directly to the hardware to receive unexpected messages.  The total_buffered_recv is intended to assist in the latter case.

I have not reviewed the code changes in this PR, so I can't say what should be done.  I was just trying to respond to the question.

File: prov/verbs/src/verbs_info.c (hunk 2 above)
Comment (shefty) [inline prov/verbs/src/verbs_info.c]: So, I think the ask for the PR is to revert the change to fi_ibv_check_rx_attr above, then it should be ready for merging.

**Comment by shefty** [comment]
I think there may still be confusion over the use of RM_ENABLED and buffered_recv.  While it's true that RM_ENABLED=1 should ensure that all sent messages eventually be received, receive side buffering is also a performance optimization.  So it can be set independent of RM_ENABLED.  I'm not sure that the checks are entirely correct, but the overall patch appears to be an improvement.  So I'm going to merge this.  We can improve the attribute checking in a separate patch.
